File: sdk/symbol_dart_sdk/bin/crypto/ed25519.dart
import './tweetNacl.dart';
import '../CryptoTypes.dart';
import '../symbol/KeyPair.dart';

const CRYPTO_SIGN_BYTES = 64;
const CRYPTO_SIGN_PUBLICKEYBYTES = 32;


/* 
class Ed25519 {
  static KeyPair keyPairFromSeed(HashMode hashMode, List<int> seed) {
    final publicKey = List<int>.filled(CRYPTO_SIGN_PUBLICKEYBYTES, 0);
    cryptoSignKeypair(hashMode[HashMode], seed, publicKey);
    return KeyPair(publicKey: publicKey, privateKey: seed);
  }

  static List<int> sign(HashMode hashMode, List<int> message, List<int> privateKey) {
    final signature = List<int>.filled(CRYPTO_SIGN_BYTES, 0);
    cryptoPrivateSign(hashMode[HashMode], privateKey, message, signature);
    return signature;
  }

  static bool verify(HashMode hashMode, List<int> message, List<int> signature, List<int> publicKey) {
    return cryptoPrivateVerify(hashMode[HashMode], publicKey, message, signature);
  }
} */

enum HashMode {
	/// Keccak hash.
	Keccak,

	/// SHA2 hash.
	Sha2_512
}

File: sdk/typescript/src/impl/ed25519.ts
// this file contains implementation details and is not intended to be used directly
import nacl from 'tweetnacl';

const crypto_sign_keypair = (seed: Uint8Array) => {
	const keyPair = nacl.sign.keyPair.fromSeed(seed);
	return keyPair;
}

const crypto_private_sign = (message: Uint8Array, privateKey: Uint8Array) => {
	const keyPair = crypto_sign_keypair(privateKey);
	const secretKey = new Uint8Array(64);
  secretKey.set(privateKey);
  secretKey.set(keyPair.publicKey, 32);
	const signature = nacl.sign.detached(message, secretKey);
	return signature;
}

const crypto_private_verify = (data: Uint8Array, signature: Uint8Array, publicKey: Uint8Array) => {
	return nacl.sign.detached.verify(data, signature, publicKey);
}

const ed25519 = {
	keyPairFromSeed: (seed: Uint8Array) => {
		const publicKey = nacl.sign.keyPair.fromSeed(seed).publicKey;	
		return { publicKey, privateKey: seed };
	},
	sign: (message: Uint8Array, privateKey: Uint8Array) => {
		const signature = crypto_private_sign(message, privateKey);
		return signature;
	},
	verify: (message: Uint8Array, signature: Uint8Array, publicKey: Uint8Array) => crypto_private_verify(message, signature, publicKey)
};

export default ed25519;
